### PR TITLE
New package: py-pyfr

### PR DESCRIPTION
**Pull request by MichaelLaufer**
Initial support for PyFR.
@FreddieWitherden

### CODE REVIEW

File: var/spack/repos/builtin/packages/py-gimmik/package.py
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-gimmik/package.py]:
```suggestion
    high performance matrix multiplication kernel code for various accelerator platforms."""
```
You can wrap this onto another line if it's too long

File: var/spack/repos/builtin/packages/py-gimmik/package.py
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-gimmik/package.py:23]:
```python
depends_on('python@3.8:', type=('build', 'run'))
```

File: var/spack/repos/builtin/packages/py-gimmik/package.py
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-gimmik/package.py]:
```suggestion
    depends_on("py-numpy@1.7:", type=('build', 'run'))
    depends_on("py-mako", type=('build', 'run'))
```

File: var/spack/repos/builtin/packages/py-pyfr/package.py
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-pyfr/package.py]:
```suggestion
    problems on streaming architectures using the Flux Reconstruction approach of Huynh."""
```

File: var/spack/repos/builtin/packages/py-pyfr/package.py
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-pyfr/package.py:36]:
```suggestion
    depends_on("python@3.8:", type=("build", "run"))
```

File: var/spack/repos/builtin/packages/py-pyfr/package.py
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-pyfr/package.py]:
```suggestion
    depends_on("py-gimmik@2.2:2")
```

File: var/spack/repos/builtin/packages/py-pyfr/package.py
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-pyfr/package.py]: I don't see scipy listed as a dep, is it actually required?

Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-pyfr/package.py]: These should all be `type=('build', 'run')`

**Review by adamjstewart — CHANGES_REQUESTED** [review]

File: var/spack/repos/builtin/packages/py-pyfr/package.py
Comment (FreddieWitherden) [reply var/spack/repos/builtin/packages/py-pyfr/package.py]: It is used opportunistically to reduce start-up time under certain circumstances.

File: var/spack/repos/builtin/packages/py-pyfr/package.py
Comment (MichaelLaufer) [reply var/spack/repos/builtin/packages/py-pyfr/package.py]: @adamjstewart why is this needed? Thanks

Comment (adamjstewart) [reply var/spack/repos/builtin/packages/py-pyfr/package.py]: The `setup.py` lists the dependency as `gimmik ~= 2.2`, which translates to `>= 2.2, == 2.*`, which translates to `@2.2:2` in Spack. See https://stackoverflow.com/a/39590286/5828163.

File: var/spack/repos/builtin/packages/py-pyfr/package.py
Comment (MichaelLaufer) [reply var/spack/repos/builtin/packages/py-pyfr/package.py:36]: According to the PyFR documentation, Python 3.9+ is needed. @FreddieWitherden.

File: var/spack/repos/builtin/packages/py-pyfr/package.py
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-pyfr/package.py]: Not sure if these should be build/run or build/link (the default). build/run means to add `prefix.bin` to `PATH` during build-time and run-time, link means that the packages links to libraries/headers.

File: var/spack/repos/builtin/packages/py-pyfr/package.py
Comment (FreddieWitherden) [reply var/spack/repos/builtin/packages/py-pyfr/package.py:36]: 3.9 as of commit `b9037dc8edf254193f9e5ba9b7706ba91cb63d2e` (which forms part of v1.13.0) since we use the dict union operators.  Setup.py seems to be only half updated (`classifiers` is correct, `python_requires` needs a bump).

File: var/spack/repos/builtin/packages/py-pyfr/package.py
Comment (FreddieWitherden) [reply var/spack/repos/builtin/packages/py-pyfr/package.py]: Is there a specific package for HIP?

Comment (adamjstewart) [reply var/spack/repos/builtin/packages/py-pyfr/package.py]: @amd-toolchain-support might know the answer to this question

Comment (MichaelLaufer) [reply var/spack/repos/builtin/packages/py-pyfr/package.py]: @FreddieWitherden, there is a package for [HIP](https://github.com/spack/spack/blob/develop/var/spack/repos/builtin/packages/hip/package.py) but it gets pulled in implicitly by the rocblas package. Can you comment on the build/link/run requirement for metis, scotch, cuda and rocblas? Thanks

File: var/spack/repos/builtin/packages/py-pyfr/package.py
Comment (MichaelLaufer) [reply var/spack/repos/builtin/packages/py-pyfr/package.py]: Thanks

File: var/spack/repos/builtin/packages/py-pyfr/package.py
Comment (MichaelLaufer) [reply var/spack/repos/builtin/packages/py-pyfr/package.py]: After some testing, and inspection of the PyFR source code, it seems that PyFR uses these opportunistically at run time, such that ` type=('run'))` is all that is needed.

**Review by adamjstewart — DISMISSED** [review]

**Comment by MichaelLaufer** [comment]
Please do not merge yet, there is some issue with the hip variant. I will debug this and update. @adamjstewart

**Comment by MichaelLaufer** [comment]
@adamjstewart I am getting unsatisfiable constraint errors when using the hip variant. It it related to the Python 3.9 requirement in PyFR  and constraints by the mesa18 package that limits Python to < 3.9. Related to #28219, #28345. Is the Python version constraint (from #28345, @eugeneswalker ) no longer needed?

**Comment by MichaelLaufer** [comment]
@adamjstewart I tried altering the mesa18 package.py to allow Python 3.9, but build issues were still encountered. Personally I think we can merge at this point as I expect this issue to be ironed out.  Your call of course....

**Review by adamjstewart — APPROVED** [review]

**Comment by adamjstewart** [comment]
Which package actually depends on `mesa18`? We should port that package to work with `mesa`. Honestly, `mesa18` and `mesa` should be combined into a single package like every other package with two build systems.

Another option would be to add Python 3.8 support to `py-pyfr`. There aren't too many new language features in Python 3.9.

**Comment by MichaelLaufer** [comment]
> Which package actually depends on `mesa18`? We should port that package to work with `mesa`. Honestly, `mesa18` and `mesa` should be combined into a single package like every other package with two build systems.
> 
> Another option would be to add Python 3.8 support to `py-pyfr`. There aren't too many new language features in Python 3.9.

Looks like @srekolam has already submitted a PR (#28491) to use the normal `mesa` package in place of `mesa18` for `hip.` I assume that that would solve the issues seen here with the PyFR hip variant.

**Comment by adamjstewart** [comment]
Merged.

**Pull request merged**
